### PR TITLE
Fix more cppcheck warnings

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR fixes some more cppcheck warnings.

**To test:**
Code review

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/ExperimentInfo.cpp [hunk 1]
```diff
@@ -79,6 +79,15 @@ ExperimentInfo::ExperimentInfo(const ExperimentInfo &source) {
   setSpectrumDefinitions(source.spectrumInfo().sharedSpectrumDefinitions());
 }
 
+/**
+ * Implements the copy assignment operator
+ */
+ExperimentInfo &ExperimentInfo::operator=(const ExperimentInfo &source) {
```
Comment (martyngigg) [inline Framework/API/src/ExperimentInfo.cpp:85]: May I suggest that we use the ["copy-and-swap"](https://stackoverflow.com/a/3652138) idiom to implement the copy constructor/assignment operator pair here? It is a nice idiom for reducing code duplication and providing exception guarantees at the same time.

File: Framework/API/src/MDGeometry.cpp [hunk 2]
```diff
@@ -102,6 +102,50 @@ MDGeometry::MDGeometry(const MDGeometry &other)
     this->setOriginalWorkspace(other.m_originalWorkspaces[i], i);
 }
 
+MDGeometry &MDGeometry::operator=(const MDGeometry &other) {
```
Comment (martyngigg) [inline Framework/API/src/MDGeometry.cpp:105]: Is this duplicated code from the copy constructor? If we can't use copy-and-swap then can we have a common method to do this?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Just a couple of notes so far on this one.

As it's quite large I'll do it in chunks.

File: Framework/DataHandling/src/Load.cpp [hunk 3]
```diff
@@ -38,12 +38,7 @@ namespace {
  * @returns true if there is exactly one string, else false.
  */
 bool isSingleFile(const std::vector<std::vector<std::string>> &fileNames) {
-  if (fileNames.size() == 1) {
-    auto first = fileNames.cbegin();
-    if (first->size() == 1)
-      return true;
-  }
-  return false;
+  return fileNames.size() == 1 ? fileNames[0u].size() == 1 : false;
```
Comment (martyngigg) [inline Framework/DataHandling/src/Load.cpp:41]: Would `return fileNames.size() == 1 && fileNames[0].size() == 1` be slightly easier to read?

File: Framework/DataHandling/src/LoadPSIMuonBin.cpp [hunk 4]
```diff
@@ -737,9 +737,9 @@ void LoadPSIMuonBin::readInTemperatureFileHeader(const std::string &contents) {
   std::string line = "";
   for (const auto charecter : contents) {
     if (charecter == '\n') {
-      if (line[0] == '!' && lineNo > uselessLines) {
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadPSIMuonBin.cpp:740]: Would this read more easily if we did the `empty` check first & once then continued?

**Review by martyngigg — COMMENTED** [review]
I'm just over a third of the way through again but I thought it might be useful to have the comments so far.

File: Framework/DataHandling/src/Load.cpp (hunk 3 above)
Comment (robertapplin) [reply Framework/DataHandling/src/Load.cpp]: Ah, that would be much better!

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 5]
```diff
@@ -1616,10 +1616,12 @@ double FitPeaks::fitFunctionHighBackground(const IAlgorithm_sptr &fit, const std
                                            const size_t &ws_index, const double &expected_peak_center,
                                            bool observe_peak_shape, const API::IPeakFunction_sptr &peakfunction,
                                            const API::IBackgroundFunction_sptr &bkgdfunc) {
+  if (!m_linearBackgroundFunction)
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply Framework/Algorithms/src/FitPeaks.cpp:1614]: Yes a nullptr could have been passed to this function, but there is a check higher up relating to the `m_highBackground` which prevents this in the current use cases.

I've just noticed I can probably just create a linear background if its missing! So I've removed the runtime error as its no longer needed

File: Framework/DataObjects/src/AffineMatrixParameter.cpp [hunk 6]
```diff
@@ -128,6 +128,12 @@ AffineMatrixParameter &AffineMatrixParameter::operator=(const AffineMatrixParame
   if (this != &other) {
     this->m_affineMatrix = other.m_affineMatrix;
     this->m_isValid = other.m_isValid;
+    size_t nx = m_affineMatrix.numRows();
```
Comment (martyngigg) [inline Framework/DataObjects/src/AffineMatrixParameter.cpp:131]: Just noting down what was discussed in person. We shouldn't need to reallocate here as the lhs/rhs are the same size. We can leave the cppcheck warning as a false positive in this case.

File: Framework/MDAlgorithms/src/LoadSQW.cpp [hunk 7]
```diff
@@ -418,25 +418,25 @@ void LoadSQW::readDNDDimensions(std::vector<Mantid::Geometry::MDHistoDimensionBu
 
   // TODO: how to use it in our framework? -> it is B^-1 matrix possibly
   // re-scaled
-  std::vector<double> u_to_Rlu(4 * 4); // the matrix transforming from lab to crystal frame with scaling
+  // std::vector<double> u_to_Rlu(4 * 4); // the matrix transforming from lab to crystal frame with scaling
```
Comment (martyngigg) [inline Framework/MDAlgorithms/src/LoadSQW.cpp:421]: Can we remove the commented out code?

File: Framework/DataHandling/src/LoadPSIMuonBin.cpp [hunk 8]
```diff
@@ -737,10 +737,10 @@ void LoadPSIMuonBin::readInTemperatureFileHeader(const std::string &contents) {
...
-        processHeaderLine(line);
-      } else if (line[0] != '!') {
+      if (line.empty() || line[0] != '!') {
         return;
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply Framework/DataHandling/src/LoadPSIMuonBin.cpp:741]: The previous behaviour was to `return`. I assume this is the right thing to do as it looks like it is trying to read the temperature from the header of a file, and the header ends when the first character is not equal to '!'

**Review by martyngigg — CHANGES_REQUESTED** [review]
Good work with this.
We're nearly there! Just a final few comments

File: Framework/DataHandling/src/LoadPSIMuonBin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/DataHandling/src/LoadPSIMuonBin.cpp:741]: Ah yes I see. Thanks for clarifying that.

**Comment by robertapplin** [comment]
> Good work with this. We're nearly there! Just a final few comments

Thanks, should be ready for another look now

**Review by martyngigg — APPROVED** [review]
Thanks for making all these extra changes. All looks good.

**Pull request merged**